sapi_demo/src/Plugin/StatisticsPlugin/ArticleColorTracker.php

```diff
@@ -0,0 +1,203 @@
+use Drupal\sapi\Plugin\StatisticsPluginBase;
+use Drupal\sapi\StatisticsItemInterface;
+
+/**
+ * This is a SAPI handler plugin used to track node views, and to count
+ * colour frequencies for the user/date, by looking at the node->field_colours
+ * value.
+ *
+ * @StatisticsPlugin(
+ *  id = "article_color_tracker",
+ *  label = "Track article colour views"
+ * )
+ *
+ * @TODO currently route-id, item-action, and colour-field-name and logger id
+ *   are all hardcoded strings.  These should be replaced or configured.
+ */
+class ArticleColorTracker extends StatisticsPluginBase implements StatisticsPluginInterface, ContainerFactoryPluginInterface {
+
+  /**
+   * Route match object as we only track on certain routes
+   *
+   * @protected \Drupal\Core\Routing\RouteMatchInterface $routeMatch
+   */
+  protected $routeMatch;
+
+  /**
+   * EntityTypeManager used to get entity storage for sapi_data items, which is
+   * used to create and edit sapi_data items from tracking.
+   *
+   * @protected Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Current user for tracking
+   *
+   * @protected Drupal\Core\Session\AccountProxyInterface $currentUser;
+   */
+  protected $currentUser;
+
+  /**
+   * ArticleColorTracker constructor.
+   *
+   * @param array $configuration
+   * @param string $plugin_id
+   * @param mixed $plugin_definition
+   * @param \Drupal\Core\Routing\RouteMatchInterface $routeMatch
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   * @param \Drupal\Core\Session\AccountProxyInterface $currentUser
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, RouteMatchInterface $routeMatch, EntityTypeManagerInterface $entityTypeManager, AccountProxyInterface $currentUser) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    $this->routeMatch = $routeMatch;
+    $this->entityTypeManager = $entityTypeManager;
+    $this->currentUser = $currentUser;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+
+      $container->get('current_route_match'),
+      $container->get('entity_type.manager'),
+      $container->get('current_user')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function process(StatisticsItemInterface $item){
+
+    // We limit this action to only listening to one type of item action
+    if ($item->getAction() != 'controller_view') {
+      return;
+    }
+
+    // we will listen only to node canonical requests
+    if ($this->routeMatch->getRouteName()!='entity.node.canonical') {
+      return;
+    }
+
+    // we will not track anonymous user
+    if ($this->currentUser->isAnonymous()) {
+      return;
+    }
```

Comment:
I will merge these three if statements I think.  They are pretty straight forward, now that the services are already available.